include/sparrow/typed_array.hpp

```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or mplied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <type_traits>
+
+#include "sparrow/array_data.hpp"
+#include "sparrow/variable_size_binary_layout.hpp"
+#include "sparrow/fixed_size_layout.hpp"
+
+namespace sparrow{
+
+template<typename T>
+struct get_default_layout {
+    static constexpr auto get_type() {
+        if constexpr (std::is_arithmetic_v<T> || std::is_array_v<T>) {
+            return std::type_identity<sparrow::fixed_size_layout<T>>{};
+        } else {
+            return std::type_identity<sparrow::variable_size_binary_layout<T, T&, const T&>>{};
+        }
+    }
+
+    using type =  typename decltype(get_type())::type;
+};
+
+// Template alias for easier usage.
+template<typename T>
+using get_default_layout_t = typename get_default_layout<T>::type;
+
+template <class T, class L = get_default_layout_t<T>>
+class typed_array
```

Comment:
@JohanMabille mentioned in https://github.com/xtensor-stack/sparrow/issues/61#issuecomment-2039177368 that the API needs to somewhat be similar to `std::vector`, and expose other methods, but I do not know what he has in mind exactly.